### DISCUSSION TITLE
Vault is empty after database restore on backup system.

### DISCUSSION THREAD

**Discussion by alexander-beck** [dani-garcia/vaultwarden#3482] [Q&A]
### Subject of the issue

Running the restore procedure leads to an empty vault of my active user sessions when the system is back up.

### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.28.1
* Web-vault version: v2023.3.0b
* OS/Arch: linux/x86_64
* Running within Docker: true (Base: Debian)
* Environment settings overridden: true
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Browser/Server Time Check: true
* Server/NTP Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: SQLite
* Database version: 3.39.2
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** SIGNUPS_DOMAINS_WHITELIST, ADMIN_TOKEN

```json
{
  "_duo_akey": null,
  "_enable_duo": false,
  "_enable_email_2fa": false,
  "_enable_smtp": true,
  "_enable_yubico": true,
  "_icon_service_csp": "",
  "_icon_service_url": "",
  "_ip_header_enabled": true,
  "_smtp_img_src": "cid:",
  "admin_ratelimit_max_burst": 3,
  "admin_ratelimit_seconds": 300,
  "admin_session_lifetime": 20,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_conn_init": "",
  "database_max_conns": 10,
  "database_timeout": 30,
  "database_url": "***************",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://*****************************",
  "domain_origin": "*****://*****************************",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "emergency_access_allowed": true,
  "emergency_notification_reminder_schedule": "0 3 * * * *",
  "emergency_request_timeout_schedule": "0 7 * * * *",
  "enable_db_wal": true,
  "event_cleanup_schedule": "0 10 0 * * *",
  "events_days_retain": null,
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "invitation_expiration_hours": 120,
  "invitation_org_name": "KNM",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": null,
  "log_level": "Info",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "login_ratelimit_max_burst": 10,
  "login_ratelimit_seconds": 60,
  "org_attachment_limit": null,
  "org_creation_users": "",
  "org_events_enabled": false,
  "org_groups_enabled": false,
  "password_hints_allowed": true,
  "password_iterations": 600000,
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sendmail_command": null,
  "sends_allowed": true,
  "sends_folder": "data/sends",
  "show_password_hint": false,
  "signups_allowed": true,
  "signups_domains_whitelist": "******",
  "signups_verify": false,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": null,
  "smtp_debug": false,
  "smtp_embed_images": true,
  "smtp_explicit_tls": null,
  "smtp_from": "",
  "smtp_from_name": "Vaultwarden",
  "smtp_host": null,
  "smtp_password": null,
  "smtp_port": 587,
  "smtp_security": "starttls",
  "smtp_ssl": null,
  "smtp_timeout": 15,
  "smtp_username": null,
  "templates_folder": "data/templates",
  "tmp_folder": "data/tmp",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_sendmail": false,
  "use_syslog": false,
  "user_attachment_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": false,
  "websocket_port": 3012,
  "yubico_client_id": null,
  "yubico_secret_key": null,
  "yubico_server": null
}
```
</details>

### Steps to reproduce

1. Create a backup on my main system.
2. Restore Database  and data folder on the backup system.
3. Assert your rsa keys on the backup system are in sync with the keys on the main system.
4. Log in on backup system (via browser)
5. Assert your vault on the backup system is in sync with the main system.
6. Stop the backup instance.
7. Restore the database and data folder on the backup system with the same backup.
8. Assert your rsa keys on the backup system are in sync with the keys on the main system.
9. Restart the backup instance 
10. Hit F5 in your browser
11. Unlock your vault.

### Expected behaviour
After unlocking my vault I should see all the data.

### Actual behaviour
After unlocking my account with my master password my vault is empty.
When I completely logout my session and login again everything is back.

**Comment by BlackDex** [comment]
If you restore a backup, you need to force all clients to resync, because there login/refresh tokens are not valid.
Also, are there any errors or logs during the unlock at either Vaultwarden or the Browser (F12 Dev Console)?

**Comment by alexander-beck** [comment]
Hi @BlackDex!

While unlocking the account I am getting a 400 error code on /identity/connect/token. The **client_id** is undefined for some reason.
I have not set up the websocket connection properly. But that should not have to do anything with this error. Should it ?

Request
`grant_type=refresh_token&client_id=undefined&refresh_token=_some-token_`

Response
```
{
    "ErrorModel": {
        "Message": "Invalid refresh token",
        "Object": "error"
    },
    "ExceptionMessage": null,
    "ExceptionStackTrace": null,
    "InnerExceptionMessage": null,
    "Message": "Invalid refresh token",
    "Object": "error",
    "ValidationErrors": {
        "": [
            "Invalid refresh token"
        ]
    },
    "error": "",
    "error_description": ""
}
```

On the server I can see the same error as well

```
[2023-04-27 06:26:18.715][request][INFO] POST /identity/connect/token
[2023-04-27 06:26:18.716][rocket::form::parser::_][TRACE] url-encoded field: ("grant_type", "refresh_token")

<details>
<summary>Details</summary>

**[2023-04-27 06:26:18.716][rocket::form::parser::_][TRACE] url-encoded field: ("client_id", "undefined")**

</details>

[2023-04-27 06:26:18.716][rocket::form::parser::_][TRACE] url-encoded field: ("refresh_token", "some-token")
[2023-04-27 06:26:18.718][response][INFO] (login) POST /identity/connect/token => 400 Bad Request
```

**Reply by BlackDex** [reply]
That is what i suspected. The refresh token is invalid because your backup database doesn't have the correct/new one. Therefore we can't verify the authentication is valid and that causes the 400.

So this is expected behavior. You should click on the force resync button in the admin interface to make sure the clients get the correct data.
